src/components/lab/canvas/molecules/CanvasNitrousOxide.js

```diff
@@ -1,94 +1,79 @@
   OXYGEN,
   NITROGEN,
-  CANVAS_MOLECULE_AREA_Y_POSITION,
   CANVAS_MOLECULES_DISTANCE_BETWEEN_VERTICAL_ATOMS,
   NEGATIVE_CHARGE,
   POSITIVE_CHARGE,
-  CANVAS_NITROUS_OXIDE_OSCILLATION_AMPLITUDE,
-  INITIAL_LINE_POINTS,
+  SPECTRUMS,
+  INTERVALS_TO_REACH_MOLECULE_CENTER,
+  INFRARED_RADIATION_PERIOD,
+  Y_SHIFT_PER_INTERVAL,
+  CANVAS_NITROUS_OXIDE_OSCILLATION_AMPLITUDES,
 } from '../../../../config/constants';
 
-const CanvasNitrousOxide = ({ x, shouldOscillate }) => {
+const CanvasNitrousOxide = ({ x, y }) => {
+  const intervalCount = useSelector(({ lab }) => lab.intervalCount);
   const spectrum = useSelector(({ lab }) => lab.spectrum);
-  const emittedLines = useSelector(({ lab }) => lab.emittedLines);
 
-  // constants to determine initial center points of atoms in this molecule when component mounts
-  // the 'initial center points' are also used in the useEffect hook below, to reset an oscillating molecule
+  // variables for determining center points of atoms in this molecule
+  // we know that after INTERVALS_TO_REACH_MOLECULE_CENTER, the radiation lines have reached the center of the molecule
+  // at this point, if the spectrum is INFRARED, the molecule should begin oscillating
+  // hence, after INTERVALS_TO_REACH_MOLECULE_CENTER, the x position of the molecule is moved every interval
   const oxygenAtomRadius = CANVAS_ATOM_DIMENSIONS[OXYGEN.size];
   const nitrogenAtomRadius = CANVAS_ATOM_DIMENSIONS[NITROGEN.size];
-  const topNitrogenAtomInitialCenterPoint = {
-    x,
+  const topNitrogenAtomCenterPoint = {
+    x:
+      intervalCount > INTERVALS_TO_REACH_MOLECULE_CENTER &&
+      spectrum === SPECTRUMS.INFRARED
+        ? x +
+          CANVAS_NITROUS_OXIDE_OSCILLATION_AMPLITUDES.TOP_NITROGEN_ATOM *
+            Math.sin(
+              intervalCount * Y_SHIFT_PER_INTERVAL * INFRARED_RADIATION_PERIOD,
+            )
+        : x,
     y:
-      CANVAS_MOLECULE_AREA_Y_POSITION -
+      y -
       nitrogenAtomRadius -
       CANVAS_MOLECULES_DISTANCE_BETWEEN_VERTICAL_ATOMS -
       nitrogenAtomRadius,
   };
-  const middleNitrogenAtomInitialCenterPoint = {
-    x,
-    y: CANVAS_MOLECULE_AREA_Y_POSITION,
+  const middleNitrogenAtomCenterPoint = {
+    x:
+      intervalCount > INTERVALS_TO_REACH_MOLECULE_CENTER &&
+      spectrum === SPECTRUMS.INFRARED
+        ? x +
+          CANVAS_NITROUS_OXIDE_OSCILLATION_AMPLITUDES.MIDDLE_NITROGEN_ATOM *
+            Math.sin(
+              intervalCount * Y_SHIFT_PER_INTERVAL * INFRARED_RADIATION_PERIOD,
+            )
+        : x,
+    y,
   };
-  const bottomOxygenAtomInitialCenterPoint = {
-    x,
+  const bottomOxygenAtomCenterPoint = {
+    x:
+      intervalCount > INTERVALS_TO_REACH_MOLECULE_CENTER &&
+      spectrum === SPECTRUMS.INFRARED
+        ? x +
+          CANVAS_NITROUS_OXIDE_OSCILLATION_AMPLITUDES.BOTTOM_OXYGEN_ATOM *
+            Math.sin(
+              intervalCount * Y_SHIFT_PER_INTERVAL * INFRARED_RADIATION_PERIOD,
+            )
+        : x,
```

Comment:
This looks similar in a lot of files. Do you think you can factor it out? It would also avoid you to change every file when a little change happens.

Reply:
You're right. This was sort of in the back of my mind while writing it. Let me look at it separately. Maybe I can pass it as a prop into these components, so I only write it one place.